File: yppm/resources/question_and_answer_community/front_end/src/generated_yrpc/question_and_answer_rpc.ts
import * as question_and_answer_objects from './question_and_answer_objects'

export class Client_question_and_answer {
  /**
   * @param {string} _service_url is something like: "http://127.0.0.1:80" or "https://127.0.0.1"
   * @param {{ [key: string]: string }} _header  http headers, it's a dictionary, liek {'content-type', 'application/json'}
   * @param {Function} _error_handle_function will get called when http request got error, you need to give it a function like: (err: String) {print(err)}
   * @param {Function} _interceptor_function will get called for every response, you need to give it a function like: (data: dict[Any, Any]) {print(data)}
   * @param {Function} _function_before_request will get called before every request, you need to give it a function like: () {global_loading_animation = true}
   * @param {Function} _function_after_request will get called after every request, you need to give it a function like: () {global_loading_animation = false}
   */
    _service_url: string
    _header: { [key: string]: string } = {}
    _error_handle_function: (error: string) => void = (error: string) => {console.log(error)}
    _special_error_key: string = "__yingshaoxo's_error__"
    _interceptor_function: (data: any) => void = (data: any) => {console.log(data)}
    _function_before_request: () => void = () => {}
    _function_after_request: () => void = () => {}

    constructor(service_url: string, header?: { [key: string]: string }, error_handle_function?: (error: string) => void, interceptor_function?: (data: any) => void, function_before_request?: () => void, function_after_request?: () => void) {
        if (service_url.endsWith("/")) {
            service_url = service_url.slice(0, service_url.length-1);
        }
        try {
            if (location.protocol === 'https:') {
                if (service_url.startsWith("http:")) {
                    service_url = service_url.replace("http:", "https:")
                }
            } else if (location.protocol === 'http:') {
                if (service_url.startsWith("https:")) {
                    service_url = service_url.replace("https:", "http:")
                }
            }
        } catch (e) {
        }
        this._service_url = service_url
        
        if (header != null) {
            this._header = header
        }

        if (error_handle_function != null) {
            this._error_handle_function = error_handle_function
        }

        if (interceptor_function != null) {
            this._interceptor_function = interceptor_function
        }

        if (function_before_request != null) {
            this._function_before_request = function_before_request
        }

        if (function_after_request != null) {
            this._function_after_request = function_after_request
        }
    } 

    async _get_reponse_or_error_by_url_path_and_input(sub_url: string, input_dict: { [key: string]: any }): Promise<any> {
        let the_url = `${this._service_url}/question_and_answer/${sub_url}/`
        try {
            this._function_before_request()
            const response = await fetch(the_url, 
            {
                method: "POST",
                body: JSON.stringify(input_dict),
                headers: {
                    "Content-type": "application/json; charset=UTF-8",
                    ...this._header
                }
            });
            var json_response = await response.json()
            this._function_after_request()
            this._interceptor_function(json_response)
            return json_response
        } catch (e) {
            this._function_after_request()
            return {[this._special_error_key]: String(e)};
        }
    }

    async about(item: question_and_answer_objects.About_Request, ignore_error?: boolean): Promise<question_and_answer_objects.About_Response | null> {
        let result = await this._get_reponse_or_error_by_url_path_and_input("about", item.to_dict())
        if (Object.keys(result).includes(this._special_error_key)) {
            if ((ignore_error == null) || ((ignore_error != null) && (!ignore_error))) {
                this._error_handle_function(result[this._special_error_key])
            }
            return null
        } else {
            return new question_and_answer_objects.About_Response().from_dict(result)
        }
    }

    async ask_yingshaoxo_ai(item: question_and_answer_objects.Ask_Yingshaoxo_Ai_Request, ignore_error?: boolean): Promise<question_and_answer_objects.Ask_Yingshaoxo_Ai_Response | null> {
        let result = await this._get_reponse_or_error_by_url_path_and_input("ask_yingshaoxo_ai", item.to_dict())
        if (Object.keys(result).includes(this._special_error_key)) {
            if ((ignore_error == null) || ((ignore_error != null) && (!ignore_error))) {
                this._error_handle_function(result[this._special_error_key])
            }
            return null
        } else {
            return new question_and_answer_objects.Ask_Yingshaoxo_Ai_Response().from_dict(result)
        }
    }

    async visitor_search(item: question_and_answer_objects.Search_Request, ignore_error?: boolean): Promise<question_and_answer_objects.Search_Response | null> {
        let result = await this._get_reponse_or_error_by_url_path_and_input("visitor_search", item.to_dict())
        if (Object.keys(result).includes(this._special_error_key)) {
            if ((ignore_error == null) || ((ignore_error != null) && (!ignore_error))) {
                this._error_handle_function(result[this._special_error_key])
            }
            return null
        } else {
            return new question_and_answer_objects.Search_Response().from_dict(result)
        }
    }

    async visitor_get_a_post(item: question_and_answer_objects.Get_A_Post_Request, ignore_error?: boolean): Promise<question_and_answer_objects.Get_A_Post_Response | null> {
        let result = await this._get_reponse_or_error_by_url_path_and_input("visitor_get_a_post", item.to_dict())
        if (Object.keys(result).includes(this._special_error_key)) {
            if ((ignore_error == null) || ((ignore_error != null) && (!ignore_error))) {
                this._error_handle_function(result[this._special_error_key])
            }
            return null
        } else {
            return new question_and_answer_objects.Get_A_Post_Response().from_dict(result)
        }
    }

    async visitor_get_comment_list_by_id_list(item: question_and_answer_objects.Get_Comment_List_By_Id_List_Request, ignore_error?: boolean): Promise<question_and_answer_objects.Get_Comment_List_By_Id_List_Response | null> {
        let result = await this._get_reponse_or_error_by_url_path_and_input("visitor_get_comment_list_by_id_list", item.to_dict())
        if (Object.keys(result).includes(this._special_error_key)) {
            if ((ignore_error == null) || ((ignore_error != null) && (!ignore_error))) {
                this._error_handle_function(result[this._special_error_key])
            }
            return null
        } else {
            return new question_and_answer_objects.Get_Comment_List_By_Id_List_Response().from_dict(result)
        }
    }

    async user_add_post(item: question_and_answer_objects.Add_Post_Request, ignore_error?: boolean): Promise<question_and_answer_objects.Add_Post_Response | null> {
        let result = await this._get_reponse_or_error_by_url_path_and_input("user_add_post", item.to_dict())
        if (Object.keys(result).includes(this._special_error_key)) {
            if ((ignore_error == null) || ((ignore_error != null) && (!ignore_error))) {
                this._error_handle_function(result[this._special_error_key])
            }
            return null
        } else {
            return new question_and_answer_objects.Add_Post_Response().from_dict(result)
        }
    }

    async user_comment_post(item: question_and_answer_objects.Comment_Post_Request, ignore_error?: boolean): Promise<question_and_answer_objects.Comment_Post_Response | null> {
        let result = await this._get_reponse_or_error_by_url_path_and_input("user_comment_post", item.to_dict())
        if (Object.keys(result).includes(this._special_error_key)) {
            if ((ignore_error == null) || ((ignore_error != null) && (!ignore_error))) {
                this._error_handle_function(result[this._special_error_key])
            }
            return null
        } else {
            return new question_and_answer_objects.Comment_Post_Response().from_dict(result)
        }
    }

    async user_download_backup_data(item: question_and_answer_objects.Download_Backup_Data_Request, ignore_error?: boolean): Promise<question_and_answer_objects.Download_Backup_Data_Response | null> {
        let result = await this._get_reponse_or_error_by_url_path_and_input("user_download_backup_data", item.to_dict())
        if (Object.keys(result).includes(this._special_error_key)) {
            if ((ignore_error == null) || ((ignore_error != null) && (!ignore_error))) {
                this._error_handle_function(result[this._special_error_key])
            }
            return null
        } else {
            return new question_and_answer_objects.Download_Backup_Data_Response().from_dict(result)
        }
    }

    async admin_download_backup_data(item: question_and_answer_objects.Admin_Download_Backup_Data_Request, ignore_error?: boolean): Promise<question_and_answer_objects.Admin_Download_Backup_Data_Response | null> {
        let result = await this._get_reponse_or_error_by_url_path_and_input("admin_download_backup_data", item.to_dict())
        if (Object.keys(result).includes(this._special_error_key)) {
            if ((ignore_error == null) || ((ignore_error != null) && (!ignore_error))) {
                this._error_handle_function(result[this._special_error_key])
            }
            return null
        } else {
            return new question_and_answer_objects.Admin_Download_Backup_Data_Response().from_dict(result)
        }
    }

    async admin_upload_backup_data(item: question_and_answer_objects.Admin_Upload_Backup_Data_Request, ignore_error?: boolean): Promise<question_and_answer_objects.Admin_Upload_Backup_Data_Response | null> {
        let result = await this._get_reponse_or_error_by_url_path_and_input("admin_upload_backup_data", item.to_dict())
        if (Object.keys(result).includes(this._special_error_key)) {
            if ((ignore_error == null) || ((ignore_error != null) && (!ignore_error))) {
                this._error_handle_function(result[this._special_error_key])
            }
            return null
        } else {
            return new question_and_answer_objects.Admin_Upload_Backup_Data_Response().from_dict(result)
        }
    }
}

export default Client_question_and_answer

File: yppm/resources/question_and_answer_community/back_end/generated_yrpc/question_and_answer_rpc.py
from .question_and_answer_objects import *


from fastapi import APIRouter, FastAPI, Request
from fastapi.staticfiles import StaticFiles
from starlette.responses import FileResponse
from fastapi.middleware.cors import CORSMiddleware
import uvicorn
import os


class Service_question_and_answer:
    async def about(self, headers: dict[str, str], item: About_Request) -> About_Response:
        default_response = About_Response()

        try:
            pass
        except Exception as e:
            print(f"Error: {e}")
            #default_response.error = str(e)
            #default_response.success = False

        return default_response

    async def ask_yingshaoxo_ai(self, headers: dict[str, str], item: Ask_Yingshaoxo_Ai_Request) -> Ask_Yingshaoxo_Ai_Response:
        default_response = Ask_Yingshaoxo_Ai_Response()

        try:
            pass
        except Exception as e:
            print(f"Error: {e}")
            #default_response.error = str(e)
            #default_response.success = False

        return default_response

    async def visitor_search(self, headers: dict[str, str], item: Search_Request) -> Search_Response:
        default_response = Search_Response()

        try:
            pass
        except Exception as e:
            print(f"Error: {e}")
            #default_response.error = str(e)
            #default_response.success = False

        return default_response

    async def visitor_get_a_post(self, headers: dict[str, str], item: Get_A_Post_Request) -> Get_A_Post_Response:
        default_response = Get_A_Post_Response()

        try:
            pass
        except Exception as e:
            print(f"Error: {e}")
            #default_response.error = str(e)
            #default_response.success = False

        return default_response

    async def visitor_get_comment_list_by_id_list(self, headers: dict[str, str], item: Get_Comment_List_By_Id_List_Request) -> Get_Comment_List_By_Id_List_Response:
        default_response = Get_Comment_List_By_Id_List_Response()

        try:
            pass
        except Exception as e:
            print(f"Error: {e}")
            #default_response.error = str(e)
            #default_response.success = False

        return default_response

    async def user_add_post(self, headers: dict[str, str], item: Add_Post_Request) -> Add_Post_Response:
        default_response = Add_Post_Response()

        try:
            pass
        except Exception as e:
            print(f"Error: {e}")
            #default_response.error = str(e)
            #default_response.success = False

        return default_response

    async def user_comment_post(self, headers: dict[str, str], item: Comment_Post_Request) -> Comment_Post_Response:
        default_response = Comment_Post_Response()

        try:
            pass
        except Exception as e:
            print(f"Error: {e}")
            #default_response.error = str(e)
            #default_response.success = False

        return default_response

    async def user_download_backup_data(self, headers: dict[str, str], item: Download_Backup_Data_Request) -> Download_Backup_Data_Response:
        default_response = Download_Backup_Data_Response()

        try:
            pass
        except Exception as e:
            print(f"Error: {e}")
            #default_response.error = str(e)
            #default_response.success = False

        return default_response

    async def admin_download_backup_data(self, headers: dict[str, str], item: Admin_Download_Backup_Data_Request) -> Admin_Download_Backup_Data_Response:
        default_response = Admin_Download_Backup_Data_Response()

        try:
            pass
        except Exception as e:
            print(f"Error: {e}")
            #default_response.error = str(e)
            #default_response.success = False

        return default_response

    async def admin_upload_backup_data(self, headers: dict[str, str], item: Admin_Upload_Backup_Data_Request) -> Admin_Upload_Backup_Data_Response:
        default_response = Admin_Upload_Backup_Data_Response()

        try:
            pass
        except Exception as e:
            print(f"Error: {e}")
            #default_response.error = str(e)
            #default_response.success = False

        return default_response


def init(service_instance: Any, router: Any):
    @router.post("/about/", tags=["question_and_answer"])
    async def about(request: Request, item: About_Request) -> About_Response:
        item = About_Request().from_dict(item.to_dict())
        headers = dict(request.headers.items())
        return (await service_instance.about(headers, item)).to_dict()

    @router.post("/ask_yingshaoxo_ai/", tags=["question_and_answer"])
    async def ask_yingshaoxo_ai(request: Request, item: Ask_Yingshaoxo_Ai_Request) -> Ask_Yingshaoxo_Ai_Response:
        item = Ask_Yingshaoxo_Ai_Request().from_dict(item.to_dict())
        headers = dict(request.headers.items())
        return (await service_instance.ask_yingshaoxo_ai(headers, item)).to_dict()

    @router.post("/visitor_search/", tags=["question_and_answer"])
    async def visitor_search(request: Request, item: Search_Request) -> Search_Response:
        item = Search_Request().from_dict(item.to_dict())
        headers = dict(request.headers.items())
        return (await service_instance.visitor_search(headers, item)).to_dict()

    @router.post("/visitor_get_a_post/", tags=["question_and_answer"])
    async def visitor_get_a_post(request: Request, item: Get_A_Post_Request) -> Get_A_Post_Response:
        item = Get_A_Post_Request().from_dict(item.to_dict())
        headers = dict(request.headers.items())
        return (await service_instance.visitor_get_a_post(headers, item)).to_dict()

    @router.post("/visitor_get_comment_list_by_id_list/", tags=["question_and_answer"])
    async def visitor_get_comment_list_by_id_list(request: Request, item: Get_Comment_List_By_Id_List_Request) -> Get_Comment_List_By_Id_List_Response:
        item = Get_Comment_List_By_Id_List_Request().from_dict(item.to_dict())
        headers = dict(request.headers.items())
        return (await service_instance.visitor_get_comment_list_by_id_list(headers, item)).to_dict()

    @router.post("/user_add_post/", tags=["question_and_answer"])
    async def user_add_post(request: Request, item: Add_Post_Request) -> Add_Post_Response:
        item = Add_Post_Request().from_dict(item.to_dict())
        headers = dict(request.headers.items())
        return (await service_instance.user_add_post(headers, item)).to_dict()

    @router.post("/user_comment_post/", tags=["question_and_answer"])
    async def user_comment_post(request: Request, item: Comment_Post_Request) -> Comment_Post_Response:
        item = Comment_Post_Request().from_dict(item.to_dict())
        headers = dict(request.headers.items())
        return (await service_instance.user_comment_post(headers, item)).to_dict()

    @router.post("/user_download_backup_data/", tags=["question_and_answer"])
    async def user_download_backup_data(request: Request, item: Download_Backup_Data_Request) -> Download_Backup_Data_Response:
        item = Download_Backup_Data_Request().from_dict(item.to_dict())
        headers = dict(request.headers.items())
        return (await service_instance.user_download_backup_data(headers, item)).to_dict()

    @router.post("/admin_download_backup_data/", tags=["question_and_answer"])
    async def admin_download_backup_data(request: Request, item: Admin_Download_Backup_Data_Request) -> Admin_Download_Backup_Data_Response:
        item = Admin_Download_Backup_Data_Request().from_dict(item.to_dict())
        headers = dict(request.headers.items())
        return (await service_instance.admin_download_backup_data(headers, item)).to_dict()

    @router.post("/admin_upload_backup_data/", tags=["question_and_answer"])
    async def admin_upload_backup_data(request: Request, item: Admin_Upload_Backup_Data_Request) -> Admin_Upload_Backup_Data_Response:
        item = Admin_Upload_Backup_Data_Request().from_dict(item.to_dict())
        headers = dict(request.headers.items())
        return (await service_instance.admin_upload_backup_data(headers, item)).to_dict()


def run(service_instance: Any, port: str, html_folder_path: str="", serve_html_under_which_url: str="/", only_return_app: bool = False):
    router = APIRouter()

    init(service_instance=service_instance, router=router)

    app = FastAPI()
    app.add_middleware(
        CORSMiddleware,
        allow_origins=['*'],
        allow_credentials=True,
        allow_methods=["*"],
        allow_headers=["*"],
    )
    app.include_router(
        router,
        prefix="/question_and_answer",
    )

    if (html_folder_path != ""):
        if os.path.exists(html_folder_path) and os.path.isdir(html_folder_path):
            app.mount(serve_html_under_which_url, StaticFiles(directory=html_folder_path, html = True), name="web")
            @app.get(serve_html_under_which_url, response_model=str)
            async def index_page():
                return FileResponse(os.path.join(html_folder_path, 'index.html'))
            @app.exception_handler(404) #type: ignore
            async def custom_404_handler(_, __): #type: ignore
                return FileResponse(os.path.join(html_folder_path, 'index.html'))
            print(f"The website is running at: http://127.0.0.1:{port}/")
        else:
            print(f"Error: You should give me an absolute html_folder_path than {html_folder_path}")

    if only_return_app == True:
        return app

    print(f"You can see the docs here: http://127.0.0.1:{port}/docs")
    uvicorn.run( #type: ignore
        app=app,
        host="0.0.0.0",
        port=int(port)
    )


if __name__ == "__main__":
    service_instance = Service_question_and_answer()
    run(service_instance, port="6060")